vnext/ReactUWP/Views/TouchEventHandler.cpp

```diff
@@ -427,6 +446,42 @@ bool TouchEventHandler::TagFromOriginalSource(
   return true;
 }
 
+winrt::IPropertyValue TouchEventHandler::TestHit(
+    const winrt::Collections::IVectorView<xaml::Documents::Inline> &inlines,
+    const winrt::Point &pointerPos) {
+  winrt::IPropertyValue tag(nullptr);
+  for (const auto &el : inlines) {
+    if (const auto span = el.try_as<winrt::Documents::Span>()) {
+      const auto resTag = TestHit(span.Inlines().GetView(), pointerPos);
+
+      if (resTag) {
+        return resTag;
+      }
+
+    } else if (const auto run = el.try_as<winrt::Documents::Run>()) {
+      const auto start = el.ContentStart();
+      const auto end = el.ContentEnd();
+
+      const auto startRect = start.GetCharacterRect(winrt::Documents::LogicalDirection::Forward);
+      const auto endRect = end.GetCharacterRect(winrt::Documents::LogicalDirection::Forward);
+
+      // Approximate the bounding rect (for now, don't account for text wrapping or right-to-left orientation)
```

Comment:
>// Approximate the bounding rect (for now, don't account for text wrapping or right-to-left orientation) [](start = 6, length = 104)

I'd like to see RTL handled in this change, but I don't think it's necessary to handle wrapping yet.

You can test RTL by setting direction='rtl' on your Text component.  To implement this, I think the simplest thing you could do is "normalize" startRect and endRect by ensuring startRect is to the left of endRect (swapping if necessary).

Please file an issue to handle the wrapping case.

Reply:
Filed #5151

Reply:
Attaching a video showing this works for the tree described in the PR description (caveats below):

![nested_vid](https://user-images.githubusercontent.com/12816515/84075260-b07f4a80-a988-11ea-87a8-ab340258fb21.gif)



While testing both LTR and RTL, I've noticed that adding a a newline character at the end of a line of text to simulate linebreaks might sometimes trip up this algorithm in certain scenarios, resulting in the text from line above being detected as clicked, eg:
```<Text> times{'\n'} </Text>``` works fine, but
```<Text> {'times\n'} </Text>``` causes from the line below this text to be reported as coming from this line - this happens because for the Run on this line: endRect == 'first character on the next line', Can be mitigated as outlined above, though, by moving the brackets around.

For RTL, however, the 'one line offset' problem happens consistently, with some X coordinates being even negative.

Let me know your thoughts on this; I'd still say that this change represents a good improvement over failing always.

Reply:
The newline characters will make the run wrap, which you've already opened a separate issue on.  I'm fine with checking in what you currently have as long as it is working in basic, non-wrapping scenarios (both LTR and RTL).

Reply:
So actually this is a RichEdit bug I filed in **2017** and Murray fixed it:
[Bug 2030427: ITextRange2::GetRect should be able to return the encompassing rectangle of a range](https://office.visualstudio.com/OC/_workitems/edit/2030427)

It's just that we probably haven't ingested the Office fix into Windows XAML (nor WinUI 3).

@chrisglein @marlenecota when are we taking a new richedit drop from Office? it sounds like we haven't taken one in a while and it's hurting us :)

Reply:
@kmelmon Yes, without newlines, my change works correctly in RTL mode.